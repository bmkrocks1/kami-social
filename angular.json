{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "cli": {
    "packageManager": "npm"
  },
  "newProjectRoot": "projects",
  "projects": {
    "kami-social": {
      "projectType": "application",
      "schematics": {
        "@schematics/angular:component": {
          "style": "scss"
        }
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "ks",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:application",
          "options": {
            "outputPath": "dist/kami-social",
            "index": "src/index.html",
            "browser": "src/main.ts",
            "polyfills": [
              "zone.js"
            ],
            "tsConfig": "tsconfig.app.json",
            "inlineStyleLanguage": "scss",
            "assets": [
              "src/favicon.ico",
              "src/assets"
            ],
            "styles": [
              "src/styles.scss"
            ],
            "scripts": []
          },
          "configurations": {
            "production": {
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "500kb",
                  "maximumError": "1mb"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "2kb",
                  "maximumError": "4kb"
                }
              ],
              "outputHashing": "all"
            },
            "development": {
              "optimization": false,
              "extractLicenses": false,
              "sourceMap": true
            }
          },
          "defaultConfiguration": "production"
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "configurations": {
            "production": {
              "buildTarget": "kami-social:build:production"
            },
            "development": {
              "buildTarget": "kami-social:build:development"
            }
          },
          "defaultConfiguration": "development"
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n",
          "options": {
            "buildTarget": "kami-social:build"
          }
        },
        "test": {
          "builder": "@angular-devkit/build-angular:karma",
          "options": {
            "polyfills": [
              "zone.js",
              "zone.js/testing"
            ],
            "tsConfig": "tsconfig.spec.json",
            "inlineStyleLanguage": "scss",
            "assets": [
              "src/favicon.ico",
              "src/assets"
            ],
            "styles": [
              "src/styles.scss"
            ],
            "scripts": []
          }
        }
      }
    }
  }
}
