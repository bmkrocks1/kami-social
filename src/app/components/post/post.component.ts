import { Component } from '@angular/core';

@Component({
  selector: 'ks-post',
  standalone: true,
  imports: [],
  templateUrl: './post.component.html',
  styleUrl: './post.component.scss',
})
export class PostComponent {}
