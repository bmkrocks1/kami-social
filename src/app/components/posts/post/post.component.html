@if (isLoading) {
  <div
    class="d-flex justify-content-center align-items-center"
    style="min-height: 300px"
  >
    <div class="spinner-border" role="status">
      <span class="visually-hidden">Loading...</span>
    </div>
  </div>
} @else if (post) {
  <h2>{{ post.title }}</h2>
  <p>{{ post.body }}</p>
}
