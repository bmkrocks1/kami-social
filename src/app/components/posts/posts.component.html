<h2>Posts</h2>

<form>
  <div class="mb-3 row">
    <div class="col-lg-5 col-xs-3 col-sm-auto">
      <input
        id="table-complete-search"
        type="text"
        class="form-control"
        name="searchTerm"
        placeholder="Search text"
        [(ngModel)]="service.searchTerm"
      />
    </div>
  </div>

  @if (service.loading$ | async) {
    <div
      class="d-flex justify-content-center align-items-center"
      style="min-height: 300px"
    >
      <div class="spinner-border" role="status">
        <span class="visually-hidden">Loading...</span>
      </div>
    </div>
  } @else {
    <table class="table table-bordered">
      <thead>
        <tr>
          <th scope="col" sortable="id" (sort)="onSort($event)">ID</th>
          <th scope="col" sortable="userId" (sort)="onSort($event)">User</th>
          <th scope="col" sortable="title" (sort)="onSort($event)">Title</th>
          <th scope="col" sortable="body" (sort)="onSort($event)">Body</th>
        </tr>
      </thead>
      <tbody>
        @for (post of posts$ | async; track post.id) {
          <tr>
            <td>
              <ngb-highlight
                [result]="post.id | number"
                [term]="service.searchTerm"
              />
            </td>
            <td>
              <ngb-highlight
                [result]="post.userId | number"
                [term]="service.searchTerm"
              />
            </td>
            <td>
              <ngb-highlight
                [result]="post.title"
                [term]="service.searchTerm"
              />
            </td>
            <td>
              <ngb-highlight [result]="post.body" [term]="service.searchTerm" />
            </td>
          </tr>
        } @empty {
          <tr>
            <td colspan="4" style="text-align: center">No posts found</td>
          </tr>
        }
      </tbody>
    </table>

    <div class="d-flex justify-content-between p-2">
      <ngb-pagination
        [collectionSize]="(total$ | async)!"
        [(page)]="service.page"
        [pageSize]="service.pageSize"
      ></ngb-pagination>

      <select
        class="form-select"
        style="width: auto"
        name="pageSize"
        [(ngModel)]="service.pageSize"
      >
        <option [ngValue]="10">10</option>
        <option [ngValue]="25">25</option>
        <option [ngValue]="50">50</option>
        <option [ngValue]="100">100</option>
      </select>
    </div>
  }
</form>
