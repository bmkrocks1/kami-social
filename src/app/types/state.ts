import { SortColumn, SortDirection } from '../directives/sortable.directive';

export interface State<T = any> {
  page: number;
  pageSize: number;
  searchTerm: string;
  sortColumn: SortColumn<T>;
  sortDirection: SortDirection;
}
