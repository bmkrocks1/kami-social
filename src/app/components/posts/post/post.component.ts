import { Component } from '@angular/core';
import { PostService } from '../../../services/post.service';
import { Post } from '../../../types/post';
import { ActivatedRoute } from '@angular/router';
import { catchError, finalize, of } from 'rxjs';

@Component({
  selector: 'ks-post',
  standalone: true,
  imports: [],
  templateUrl: './post.component.html',
  styleUrl: './post.component.scss',
  providers: [PostService],
})
export class PostComponent {
  post: Post | null = null;
  postId!: number;

  isLoading = false;

  constructor(
    private postService: PostService,
    private activatedRoute: ActivatedRoute
  ) {
    //
  }

  ngOnInit(): void {
    const postId = this.activatedRoute.snapshot.paramMap.get('id');
    if (postId) {
      try {
        this.postId = parseInt(postId, 10);
        this.getPost(this.postId);
      } catch (e) {
        // TODO
      }
    }
  }

  getPost(postId: number): void {
    this.isLoading = true;
    this.postService
      .getPostById(postId)
      .pipe(
        catchError((error) => {
          // TODO: handle error
          return of(error);
        }),
        finalize(() => {
          this.isLoading = false;
        })
      )
      .subscribe((post) => (this.post = post));
  }
}
