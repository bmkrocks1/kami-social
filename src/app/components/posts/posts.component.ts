import { AsyncPipe, DecimalPipe } from '@angular/common';
import { Component, OnInit, QueryList, ViewChildren } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NgbHighlight, NgbPaginationModule } from '@ng-bootstrap/ng-bootstrap';
import { Observable } from 'rxjs';

import { Post } from '../../types/post';
import { PostsService } from '../../services/posts.service';
import {
  NgbdSortableHeader,
  SortEvent,
} from '../../directives/sortable.directive';
import { ActivatedRoute, Params, Router } from '@angular/router';
import { State } from '../../types/state';

@Component({
  selector: 'ks-posts',
  standalone: true,
  imports: [
    DecimalPipe,
    FormsModule,
    AsyncPipe,
    NgbHighlight,
    NgbdSortableHeader,
    NgbPaginationModule,
  ],
  templateUrl: './posts.component.html',
  styleUrl: './posts.component.scss',
  providers: [PostsService, DecimalPipe],
})
export class PostsComponent implements OnInit {
  posts$: Observable<Post[]>;
  total$: Observable<number>;

  @ViewChildren(NgbdSortableHeader) headers!: QueryList<
    NgbdSortableHeader<Post>
  >;

  constructor(
    private router: Router,
    private activatedRoute: ActivatedRoute,
    public service: PostsService
  ) {
    this.posts$ = service.posts$;
    this.total$ = service.total$;
  }

  ngOnInit() {
    const state: State<Post> = Object.entries(
      this.activatedRoute.snapshot.queryParams
    )
      .filter(([key]) =>
        [
          'page',
          'pageSize',
          'searchTerm',
          'sortColumn',
          'sortDirection',
        ].includes(key)
      )
      .reduce((acc, [key, value]) => {
        return {
          ...acc,
          [key]: key === 'page' || key === 'pageSize' ? Number(value) : value,
        };
      }, {} as State<Post>);

    this.service.state = state;

    this.service.state$.subscribe((state) => {
      const queryParams: Params = (
        Object.entries(state) as [keyof State, any][]
      ).reduce((params, [key, value]: [keyof State, any]) => {
        if (
          (key === 'page' && value === 1) ||
          (key === 'pageSize' && value === 10) ||
          !value
        ) {
          params[key] = undefined;
          return params;
        }

        params[key] = value;
        return params;
      }, {} as Params);

      console.log('navigate:', queryParams);

      this.router.navigate([], {
        relativeTo: this.activatedRoute,
        queryParams,
        queryParamsHandling: 'merge',
      });
    });
  }

  onSort({ column, direction }: SortEvent<Post>) {
    // resetting other headers
    this.headers.forEach((header) => {
      if (header.sortable !== column) {
        header.direction = '';
      }
    });

    this.service.sortColumn = column;
    this.service.sortDirection = direction;
  }
}
