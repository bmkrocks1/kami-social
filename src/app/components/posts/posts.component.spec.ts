import { render, screen } from '@testing-library/angular';
import { PostsComponent } from './posts.component';
import { PostsService } from '../../services/posts.service';
import { HttpClientModule } from '@angular/common/http';
import { Post } from '../../types/post';
import { of } from 'rxjs';

const MOCK_POSTS: Post[] = [
  {
    userId: 1,
    id: 1,
    title: 'foo',
    body: 'foo',
  },
  {
    userId: 2,
    id: 2,
    title: 'bar',
    body: 'bar',
  },
];

const postsSpy = jest
  .spyOn(PostsService.prototype, 'posts$', 'get')
  .mockReturnValue(of(MOCK_POSTS));

const totalSpy = jest
  .spyOn(PostsService.prototype, 'total$', 'get')
  .mockReturnValue(of(MOCK_POSTS.length));

const loadingSpy = jest
  .spyOn(PostsService.prototype, 'loading$', 'get')
  .mockReturnValue(of(false));

describe('PostsComponent', () => {
  test('should render posts', async () => {
    const result = await render(PostsComponent, {
      imports: [HttpClientModule],
    });

    expect(postsSpy).toHaveBeenCalled();
    expect(totalSpy).toHaveBeenCalled();
    expect(loadingSpy).toHaveBeenCalled();

    const table: HTMLTableElement = screen.getByTestId('posts-table');
    expect(table).toBeInTheDocument();

    const rows: HTMLTableRowElement[] =
      screen.queryAllByTestId('posts-table-row');
    expect(rows.length).toBe(MOCK_POSTS.length);

    expect(result.container).toMatchSnapshot();
  });
});
