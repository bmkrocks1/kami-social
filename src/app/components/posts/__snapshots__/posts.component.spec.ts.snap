// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsComponent should render posts 1`] = `
<div
  id="root0"
>
  <h2>
    Posts
  </h2>
  <form
    class="ng-untouched ng-pristine ng-valid"
    novalidate=""
  >
    <div
      class="mb-4 row"
    >
      <div
        class="col-lg-5 col-xs-6"
      >
        <input
          class="form-control ng-untouched ng-pristine ng-valid"
          id="table-complete-search"
          name="searchTerm"
          placeholder="Search text"
          type="text"
        />
      </div>
    </div>
    <table
      class="table table-bordered"
      data-testid="posts-table"
    >
      <thead>
        <tr>
          <th
            class="bg-light"
            scope="col"
            sortable="id"
            width="5%"
          >
             ID 
          </th>
          <th
            class="bg-light"
            scope="col"
            sortable="userId"
            width="8%"
          >
             User 
          </th>
          <th
            class="bg-light"
            scope="col"
            sortable="title"
            width="37%"
          >
             Title 
          </th>
          <th
            class="bg-light"
            scope="col"
            sortable="body"
            width="60%"
          >
             Body 
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          data-testid="posts-table-row"
        >
          <td>
             1 
          </td>
          <td>
            <a
              class="text-decoration-none"
              href="/user/1"
            >
               1 
            </a>
          </td>
          <td>
            <a
              class="text-decoration-none"
              href="/1"
            >
              <ngb-highlight>
                foo
                
                
                
                
              </ngb-highlight>
            </a>
          </td>
          <td>
            <ngb-highlight>
              foo
              
              
              
              
            </ngb-highlight>
          </td>
        </tr>
        <tr
          data-testid="posts-table-row"
        >
          <td>
             2 
          </td>
          <td>
            <a
              class="text-decoration-none"
              href="/user/2"
            >
               2 
            </a>
          </td>
          <td>
            <a
              class="text-decoration-none"
              href="/2"
            >
              <ngb-highlight>
                bar
                
                
                
                
              </ngb-highlight>
            </a>
          </td>
          <td>
            <ngb-highlight>
              bar
              
              
              
              
            </ngb-highlight>
          </td>
        </tr>
        
        
        
        
      </tbody>
    </table>
    <div
      class="d-flex justify-content-between p-2 mt-4"
    >
      <ngb-pagination
        role="navigation"
      >
        
        
        
        
        
        
        
        <ul
          class="pagination"
        >
          
          <li
            class="page-item disabled"
          >
            <a
              aria-disabled="true"
              aria-label="Previous"
              class="page-link"
              href=""
              tabindex="-1"
            >
              <span
                aria-hidden="true"
              >
                «
              </span>
              
            </a>
          </li>
          
          <li
            aria-current="page"
            class="page-item active"
          >
            <a
              class="page-link"
              href=""
            >
              1
              
            </a>
            
            
          </li>
          
          
          <li
            class="page-item disabled"
          >
            <a
              aria-disabled="true"
              aria-label="Next"
              class="page-link"
              href=""
              tabindex="-1"
            >
              <span
                aria-hidden="true"
              >
                »
              </span>
              
            </a>
          </li>
          
          
        </ul>
      </ngb-pagination>
      <select
        class="form-select ng-untouched ng-pristine ng-valid"
        name="pageSize"
        style="width: auto;"
      >
        <option
          value="0: 10"
        >
          10
        </option>
        <option
          value="1: 25"
        >
          25
        </option>
        <option
          value="2: 50"
        >
          50
        </option>
        <option
          value="3: 100"
        >
          100
        </option>
      </select>
    </div>
  </form>
</div>
`;
